### PR TITLE
Update C,U,D & Add mapper to base controller

### PR DESCRIPTION
**Pull request by AnhQuoc2906**
- Currently, mapper in base controller is in no use yet. It can be use in future functions.
- UpdateUserRequest is first used in update user, but I can't use it due to some dependencies, so I update by adding single entities to the current user (which is target in UserService - Update)
- Adding CreatedAt, ModifiedAt, DeletedAt to GenericRepository

### CODE REVIEW

File: Api/UserFeatures/Requests/UpdateUserRequest.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿using DomainLayer.Entities;
+using WebApiLayer.Mappings;
+
+namespace WebApiLayer.UserFeatures.Requests;
+
+public class UpdateUserRequest : IMapTo<UserEntity>
```
Comment (danielNg91) [inline Api/UserFeatures/Requests/UpdateUserRequest.cs:6]: create class UpdateUserRequest but [UpdateUser](https://github.com/FPTU-GHub-Capstone/game-management-service/blob/199bfa55501f7adca40bd933bddbfe220d2b1534/Api/Controllers/UsersController.cs#L41C1-L41C11) use UserEntity??

File: Application/Business/AttributeGroup/AttributeGroupServices.cs [hunk 2]
```diff
@@ -22,8 +24,7 @@ public async Task<int> Count()
     {
         return await _attributeRepo.CountAsync();
     }
-    public async Task Create(AttributeGroupEntity attributeGroup)
-    {
+    public async Task Create(AttributeGroupEntity attributeGroup){
         var attList = List();
```
Comment (danielNg91) [inline Application/Business/AttributeGroup/AttributeGroupServices.cs:28]: we accept duplicated att group name in different game ==> re-check logic

File: Application/Business/AttributeGroup/AttributeGroupServices.cs [hunk 3]
```diff
@@ -36,20 +37,33 @@ public async Task Create(AttributeGroupEntity attributeGroup)
...
+        }
+        else
+        {
+            throw new BadRequestException("Name already exist.");
         }
     }
     public async Task Update(Guid attributeGroupid, AttributeGroupEntity attributeGroup)
     {
-
+        var target = await GetById(attributeGroupid);
```
Comment (danielNg91) [inline Application/Business/AttributeGroup/AttributeGroupServices.cs:49]: use this code style to make code cleaner
```
if (target == null)) {
throw...
}
...handle
return....
```

File: Application/Business/AttributeGroup/AttributeGroupServices.cs [hunk 4]
```diff
@@ -36,20 +37,33 @@ public async Task Create(AttributeGroupEntity attributeGroup)
...
+        }
+        else
+        {
+            throw new NotFoundException("Attribute group not exist");
+        }
     }
     public async Task Delete(Guid attributeGroupid)
     {
-
+        var target = await _attributeRepo.DeleteSoftAsync(attributeGroupid);
```
Comment (danielNg91) [inline Application/Business/AttributeGroup/AttributeGroupServices.cs:63]: check this function logic, why do you delete before checking if this entity exists???

File: Application/Business/UserServices/UserServices.cs [hunk 5]
```diff
@@ -36,28 +38,35 @@ public async Task<int> Count()
         }
```
Comment (danielNg91) [inline Application/Business/UserServices/UserServices.cs:38]: bad practice. rewrite this query, you should let the db do this for you

File: Application/Business/UserServices/UserServices.cs [hunk 6]
```diff
@@ -36,28 +38,35 @@ public async Task<int> Count()
...
-                Code = user.Code,
-                Status = user.Status,
-                Balance = user.Balance
-            };
             await _userRepo.CreateAsync(user);
         }
     }
     public async Task Update(Guid UserId, UserEntity user) {
         var target = await GetById(UserId);
         if(target is not null)
```
Comment (danielNg91) [inline Application/Business/UserServices/UserServices.cs:46]: use Automapper in controller and should not handle like this in service layer

**Review by danielNg91 — CHANGES_REQUESTED** [review]

File: Api/Controllers/AttributesController.cs [hunk 7]
```diff
@@ -17,33 +17,37 @@ public AttributesController(IAttributeGroupServices attributeServices, IGenericR
...
         await _attributeServices.Create(attributeGroup);
-        return NoContent();
+        return Ok();
     }
 
     [HttpPut("{id}")]
-    public void Put(int id, [FromBody] string value)
+    public async Task<IActionResult> UpdateAttributeGroup(string id, [FromBody] AttributeGroupEntity attributeGroup)
     {
+        await _attributeServices.Update(Guid.Parse(id),attributeGroup);
```
[the remark this answers is not in the document]
Comment (AnhQuoc2906) [reply Api/Controllers/AttributesController.cs:43]: Return NoContent

File: Api/Controllers/AttributesController.cs [hunk 8]
```diff
@@ -17,33 +17,37 @@ public AttributesController(IAttributeGroupServices attributeServices, IGenericR
...
+        return Ok();
     }
 
     [HttpDelete("{id}")]
-    public void Delete(int id)
+    public async Task<IActionResult> Delete(string id)
     {
+        await _attributeServices.Delete(Guid.Parse(id));
+        return Ok();
```
[the remark this answers is not in the document]
Comment (AnhQuoc2906) [reply Api/Controllers/AttributesController.cs:51]: Return NoContent

File: Api/Controllers/AttributesController.cs [hunk 9]
```diff
@@ -17,33 +17,37 @@ public AttributesController(IAttributeGroupServices attributeServices, IGenericR
...
-    public async Task<IActionResult> Get(string id)
+    public async Task<IActionResult> GetAttributeGroup(string id)
     {
         var attribute = _attributeServices.GetById(Guid.Parse(id));
         return Ok(attribute.Result);
     }
 
     [HttpPost]
-    public async Task<IActionResult> Post([FromBody] AttributeGroupEntity attributeGroup)
+    public async Task<IActionResult> CreateAttributeGroup([FromBody] AttributeGroupEntity attributeGroup)
```
[the remark this answers is not in the document]
Comment (AnhQuoc2906) [reply Api/Controllers/AttributesController.cs:34]: added

File: Application/Business/AttributeGroup/AttributeGroupServices.cs (hunk 4 above)
Comment (AnhQuoc2906) [reply Application/Business/AttributeGroup/AttributeGroupServices.cs:63]: The function DeleteSoftAsync changes DeletedAt and returns item if it exist. 

```
public virtual async Task<T> DeleteSoftAsync(Guid id)
    {
        T _entity = await FindByIdAsync(id);
        if (_entity == null)
        {
            return null;
        }
        _entity.DeletedAt = DateTime.UtcNow;
        await UpdateAsync(_entity);
        return _entity;
    }
```

File: Api/UserFeatures/Requests/UpdateUserRequest.cs (hunk 1 above)
Comment (AnhQuoc2906) [reply Api/UserFeatures/Requests/UpdateUserRequest.cs:6]: Removed

File: Application/Extensions/BadRequestException.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
+﻿using AutoWrapper.Wrappers;
```
Comment (danielNg91) [inline Application/Extensions/BadRequestException.cs:1]: these exceptions already exist in domainlayer, let reuse it

File: Api/Controllers/AttributesController.cs [hunk 11]
```diff
@@ -17,33 +18,41 @@ public AttributesController(IAttributeGroupServices attributeServices, IGenericR
...
-    public async Task<IActionResult> Get()
+    public async Task<IActionResult> GetAttributeGroups()
     {
         var attribute = await _attributeServices.List();
         return Ok(attribute);
     }
 
     [HttpGet("{id}")]
-    public async Task<IActionResult> Get(string id)
+    public async Task<IActionResult> GetAttributeGroup(string id)
```
Comment (danielNg91) [inline Api/Controllers/AttributesController.cs:28]: we use Guid as key, use Guid as type of id to take advantage of the validation feature

File: Api/Controllers/AttributesController.cs [hunk 12]
```diff
@@ -17,33 +18,41 @@ public AttributesController(IAttributeGroupServices attributeServices, IGenericR
...
     }
 
     [HttpPut("{id}")]
-    public void Put(int id, [FromBody] string value)
+    public async Task<IActionResult> UpdateAttributeGroup(string id, [FromBody] UpdateAttributeGroupRequest attributeGroup)
     {
+        AttributeGroupEntity attGrpEnt = new AttributeGroupEntity();
+        Mapper.Map(attributeGroup, attGrpEnt);
+        await _attributeServices.Update(Guid.Parse(id), attGrpEnt);
+        return NoContent();
```
Comment (danielNg91) [inline Api/Controllers/AttributesController.cs:49]: return the updated entity to help FE easier when integrate with the APIs

File: Api/Controllers/UsersController.cs [hunk 13]
```diff
@@ -31,8 +32,10 @@ public async Task<IActionResult> GetUser(string id)
     }
 
     [HttpPost]
-    public async Task<IActionResult> CreateUser([FromBody] UserEntity user)
+    public async Task<IActionResult> CreateUser([FromBody] CreateUserRequest cUser)
     {
+        UserEntity user = new UserEntity();
+        Mapper.Map(cUser, user);
         await _userServices.Create(user);
         return Ok();
```
Comment (danielNg91) [inline Api/Controllers/UsersController.cs:40]: return 201

File: Application/Business/AttributeGroup/AttributeGroupServices.cs [hunk 14]
```diff
@@ -22,34 +27,40 @@ public async Task<int> Count()
...
     {
-        var attList = List();
-        var check = false;
-        foreach (var attgrp in attList.Result)
+        var target = await GetById(attributeGroupid);
+        if(target is not null)
         {
-            if (attgrp.Name == attributeGroup.Name)
+            Type entityType = typeof(AttributeGroupEntity);
+            foreach (PropertyInfo propertyInfo in entityType.GetProperties())
```
Comment (danielNg91) [inline Application/Business/AttributeGroup/AttributeGroupServices.cs:39]: use Automapper handle this feature, prevent use the same logic in other places

File: Application/Business/AttributeGroup/AttributeGroupServices.cs [hunk 15]
```diff
@@ -22,34 +27,40 @@ public async Task<int> Count()
...
-        if (!check)
+        else
         {
-            var attGrp = new AttributeGroupEntity
-            {
-                Name = attributeGroup.Name,
-                Effect = attributeGroup.Effect
-            };
-            await _attributeRepo.CreateAsync(attGrp);
+            throw new NotFoundException("Attribute group not exist");
```
Comment (danielNg91) [inline Application/Business/AttributeGroup/AttributeGroupServices.cs:55]: let check if Attribute group not exist ==> return, then do logic to reduce nested block

File: Application/Business/AttributeGroup/AttributeGroupServices.cs (hunk 4 above)
Comment (danielNg91) [reply Application/Business/AttributeGroup/AttributeGroupServices.cs:60]: finding non-entity is user error, return error to client. You have to check whether the entity exists or not before performing the logic

File: Application/Business/UserServices/UserServices.cs [hunk 16]
```diff
@@ -25,39 +29,41 @@ public async Task<int> Count()
...
-                Status = user.Status,
-                Balance = user.Balance
-            };
-            await _userRepo.CreateAsync(user);
-        }
+        await _userRepo.CreateAsync(user);
     }
     public async Task Update(Guid UserId, UserEntity user) {
         var target = await GetById(UserId);
         if(target is not null)
```
Comment (danielNg91) [inline Application/Business/UserServices/UserServices.cs:43]: use auto mapper, remove this verbose code

File: Application/ServiceLayer.csproj [hunk 17]
```diff
@@ -13,12 +13,12 @@
 
   <ItemGroup>
     <Folder Include="Exceptions\" />
-    <Folder Include="Extensions\" />
     <Folder Include="Models\" />
     <Folder Include="Utilities\" />
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\Domain\DomainLayer.csproj" />
```
Comment (danielNg91) [inline Application/ServiceLayer.csproj:21]: Repository layer already include domain, remove this ProjectReference

**Review by danielNg91 — CHANGES_REQUESTED** [review]

File: Api/Controllers/UsersController.cs (hunk 13 above)
Comment (AnhQuoc2906) [reply Api/Controllers/UsersController.cs:40]: `return CreatedAtAction(nameof(GetUser),new {id=user.Id}, user);`

File: Api/Controllers/AttributesController.cs (hunk 12 above)
Comment (AnhQuoc2906) [reply Api/Controllers/AttributesController.cs:49]: `return AcceptedAtAction(nameof(GetAttributeGroup), new { id = Guid.Parse(id) }, attGrpEnt);`

File: Api/Controllers/AttributesController.cs (hunk 11 above)
Comment (AnhQuoc2906) [reply Api/Controllers/AttributesController.cs:28]:
```
public async Task<IActionResult> GetAttributeGroup(Guid id)
    {
        var attribute = await _attributeServices.GetById(id);
        return Ok(attribute);
    }
```

Comment (danielNg91) [reply Api/Controllers/AttributesController.cs:28]: return Ok

File: Api/Controllers/AttributesController.cs (hunk 12 above)
Comment (danielNg91) [reply Api/Controllers/AttributesController.cs:49]: return Ok

Comment (danielNg91) [reply Api/Controllers/AttributesController.cs:49]: return Ok

File: Application/Business/UserServices/UserServices.cs [hunk 18]
```diff
@@ -25,39 +29,41 @@ public async Task<int> Count()
...
     public async Task Create(UserEntity user) {
         var userList = List();
-        var check = false;
         foreach(var User in userList.Result)
         {
             if(User.Email == user.Email)
             {
-                check = true;
-                break;
+                throw new ConflictException("Email exists");
```
Comment (danielNg91) [inline Application/Business/UserServices/UserServices.cs:36]: https://developer.mozilla.org/en-US/docs/web/http/status/409

**Review by danielNg91 — CHANGES_REQUESTED** [review]

File: Application/Business/AttributeGroup/AttributeGroupServices.cs (hunk 14 above)
Comment (AnhQuoc2906) [reply Application/Business/AttributeGroup/AttributeGroupServices.cs:39]: Updating in Controller and remove logic in service:
Controller: 
```
public async Task<IActionResult> UpdateAttributeGroup(Guid id, [FromBody] UpdateAttributeGroupRequest attributeGroup)
    {
        AttributeGroupEntity attGrpEnt = await _attributeServices.GetById(id);
        Mapper.Map(attributeGroup, attGrpEnt);
        await _attributeServices.Update(id, attGrpEnt);
        return AcceptedAtAction(nameof(GetAttributeGroup), new { id = id }, attGrpEnt);
    }
```
Service:
```
public async Task Update(Guid attributeGroupid, AttributeGroupEntity attributeGroup)
    {
        var target = await GetById(attributeGroupid);
        if(target is not null)
        {
            await _attributeRepo.UpdateAsync(target);
        }
        else
        {
            throw new NotFoundException("Attribute group not exist");
        }
    }
```

**Pull request closed without merging**